# Task: Linear vs Polynomial Regression

## Objective
Build and compare Linear Regression and Polynomial Regression models then select the best performing model and interpret its learned equation.


## Dataset
Use the **Boston Housing dataset**.
    
    from sklearn.datasets import load_boston

In [16]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler


In [17]:
# 1. Load the Dataset & 2. Convert to DataFrame
try:
    # Attempt to load from sklearn (works for sklearn versions < 1.2)
    from sklearn.datasets import load_boston
    boston = load_boston()
    df = pd.DataFrame(boston.data, columns=boston.feature_names)
    df['TARGET'] = boston.target
    print("Dataset loaded successfully from scikit-learn.")
    
except ImportError:
    # Alternative method: fetch directly from the source if sklearn import fails
    print("load_boston not found in sklearn. Loading from alternative source...")
    data_url = "http://lib.stat.cmu.edu/datasets/boston"
    raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
    data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
    target = raw_df.values[1::2, 2]
    
    # Feature names for the Boston dataset
    columns = ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT']
    df = pd.DataFrame(data, columns=columns)
    df['TARGET'] = target
    print("Dataset loaded successfully from original URL.")

<>:14: SyntaxWarning: invalid escape sequence '\s'
<>:14: SyntaxWarning: invalid escape sequence '\s'
C:\Users\ahmedthabet\AppData\Local\Temp\ipykernel_28544\4094326483.py:14: SyntaxWarning: invalid escape sequence '\s'
  raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)


load_boston not found in sklearn. Loading from alternative source...
Dataset loaded successfully from original URL.


### 1. Data Preparation

In [50]:
# 3. Data Inspection & Overview
print("\n--- Dataset Info ---")
df.info()



--- Dataset Info ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    float64
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    float64
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
 13  TARGET   506 non-null    float64
dtypes: float64(14)
memory usage: 55.5 KB


In [51]:
print("\n--- Statistical Summary ---")
print(df.describe())


--- Statistical Summary ---
             CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.601545   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.677083   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534  356.674032  

In [52]:
print("\n--- Checking for missing values ---")
print(df.isnull().sum()) 
print("\nDataFrame Shape:", df.shape)


--- Checking for missing values ---
CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
TARGET     0
dtype: int64

DataFrame Shape: (506, 14)


In [55]:
# 4. Separate X and y
X = df.drop('TARGET', axis=1)
y = df['TARGET']

# 5. Train/Test Split (80% training, 20% testing)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=101)

In [56]:
# 6. Feature Scaling
scaler = StandardScaler()
# Fit on training data ONLY, then transform both
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

print("\nData Preparation Complete! X_train_scaled and X_test_scaled are ready.")


Data Preparation Complete! X_train_scaled and X_test_scaled are ready.


### 2. Linear Regression Model
- Train a Linear Regression model
- Make predictions on the test set
- Compute evaluation metrics:
  - MAE
  - MSE
  - RMSE

In [57]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error
import numpy as np

# 1. Initialize the Linear Regression model
lr_model = LinearRegression()

In [58]:
# 2. Train the model using the scaled training data
lr_model.fit(X_train_scaled, y_train)

# 3. Make predictions on the scaled test data
y_pred_lr = lr_model.predict(X_test_scaled)

In [59]:
# 4. Compute evaluation metrics
mae_lr = mean_absolute_error(y_test, y_pred_lr)
mse_lr = mean_squared_error(y_test, y_pred_lr)
rmse_lr = np.sqrt(mse_lr)



In [60]:
# Print the results
print("--- Linear Regression Model Evaluation ---")
print(f"Mean Absolute Error (MAE): {mae_lr:.4f}")
print(f"Mean Squared Error (MSE): {mse_lr:.4f}")
print(f"Root Mean Squared Error (RMSE): {rmse_lr:.4f}")

--- Linear Regression Model Evaluation ---
Mean Absolute Error (MAE): 3.9994
Mean Squared Error (MSE): 33.2689
Root Mean Squared Error (RMSE): 5.7679


### 3. Polynomial Regression Model
- Train a Polynomial Regression model
- Make predictions on the test set
- Compute evaluation metrics:
  - MAE
  - MSE
  - RMSE


In [61]:
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error
import numpy as np

degrees = [2, 3, 4]
best_degree = None
best_rmse = float('inf')
best_poly_metrics = {}

In [62]:
print("--- Testing Different Polynomial Degrees ---")

for d in degrees:
    # 1. Transform features into polynomial features for degree 'd'
    poly = PolynomialFeatures(degree=d)
    X_train_poly = poly.fit_transform(X_train_scaled)
    X_test_poly = poly.transform(X_test_scaled)
    
    # 2. Initialize and train the model
    poly_model = LinearRegression()
    poly_model.fit(X_train_poly, y_train)
    
    # 3. Make predictions
    y_pred_poly = poly_model.predict(X_test_poly)
    
    # 4. Compute metrics
    mae = mean_absolute_error(y_test, y_pred_poly)
    mse = mean_squared_error(y_test, y_pred_poly)
    rmse = np.sqrt(mse)
    
    print(f"Degree {d} -> MAE: {mae:.4f} | MSE: {mse:.4f} | RMSE: {rmse:.4f}")
    
    # 5. Track the best degree based on the lowest RMSE
    if rmse < best_rmse:
        best_rmse = rmse
        best_degree = d
        best_poly_metrics = {'MAE': mae, 'MSE': mse, 'RMSE': rmse}

--- Testing Different Polynomial Degrees ---
Degree 2 -> MAE: 2.8468 | MSE: 17.1358 | RMSE: 4.1395
Degree 3 -> MAE: 93.9490 | MSE: 60951.0230 | RMSE: 246.8826
Degree 4 -> MAE: 17.9263 | MSE: 4364.2452 | RMSE: 66.0624


In [63]:
print(f"\n>>> The best Polynomial Degree based on test RMSE is: {best_degree} <<<")


>>> The best Polynomial Degree based on test RMSE is: 2 <<<


### 4. Model Comparison
- Compare Linear vs Polynomial Regression results
- Identify which model performs better on the test data and justify your decision


In [64]:
import pandas as pd

# Creating a summary dataframe using your actual test results
results = {
    'Model': ['Linear Regression', 'Polynomial Regression (Degree 2)'],
    'MAE': [3.9994, 2.8468],
    'MSE': [33.2689, 17.1358],
    'RMSE': [5.7679, 4.1395]
}

comparison_df = pd.DataFrame(results)

print("--- Final Model Comparison ---")
print(comparison_df.to_string(index=False))

--- Final Model Comparison ---
                           Model    MAE     MSE   RMSE
               Linear Regression 3.9994 33.2689 5.7679
Polynomial Regression (Degree 2) 2.8468 17.1358 4.1395


The Polynomial Regression model (Degree 2) is the clear winner and performs significantly better on the unseen test data.

Justification:

Lower Average Error: The Mean Absolute Error (MAE) dropped from ~4.00 in the linear model to ~2.85. In the context of the dataset, this means our average prediction error improved from roughly $4,000 off the actual price to only $2,850 off.

Fewer Extreme Mistakes: The Root Mean Squared Error (RMSE) also saw a substantial decrease (from 5.77 down to 4.14). Because RMSE heavily penalizes large errors, this drop proves that the Degree 2 model makes far fewer massive miscalculations than the standard linear model.

Capturing Complexity: The standard Linear Regression model assumes a perfectly straight-line relationship between every feature (like crime rate or age) and the house price. By expanding our features to a 2nd-degree polynomial, the model was able to capture the natural curves and interactions present in the real-world housing market, leading to a much tighter fit without overfitting (which we observed happening at Degrees 3 and 4).